### PR TITLE
Add a way to register new resource pack providers.

### PR DESCRIPTION
**Pull request by LambdAurora**
As the title says.

Also contains some extra stuff like a QuiltResourcePack interface that is injected in ResourcePack, this allows to make stuff a bit more generic and expose some new stuff to API.

This is not the virtual resource pack API, but it's a step closer to it.

Use cases:
 - visible virtual resource packs, but only the registration of the resource pack profile is provided
 - global datapack directory, as it would use a resource pack provider

The testmod includes a test virtual resource pack profile, it generates a random dirt and pack.png textures.

### CODE REVIEW

File: library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/GroupResourcePack.java
Comment (TheGlitch76) [inline library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/GroupResourcePack.java]: unnecessary braces?

File: library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/QuiltResourcePack.java
Comment (TheGlitch76) [inline library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/QuiltResourcePack.java]:
```suggestion
	 * has the last word.
```

Comment (TheGlitch76) [inline library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/QuiltResourcePack.java]: it's not immediately clear what you mean by it having "the last word". Do you mean that if you set an ResourcePackProfile activation type, this value is ignored?

File: library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/QuiltResourcePack.java
Comment (TheGlitch76) [inline library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/QuiltResourcePack.java:29]: Is this meant to be ApiStatus.NonExtendable then?

File: library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/GroupResourcePack.java
Comment (LambdAurora) [reply library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/GroupResourcePack.java]: Necessary braces, this is the inline return notation, the inline return notation has specific semantics, in addition of filling the Returns field, it also generates a summary/description as `Returns <content>.`.

File: library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/QuiltResourcePack.java
Comment (LambdAurora) [reply library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/QuiltResourcePack.java]: Yeah, if your ResourcePackProfile says "always enabled", the value will be ignored.

File: library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/QuiltResourcePack.java
Comment (LambdAurora) [reply library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/QuiltResourcePack.java:29]: Meh, ResourcePack extends it via interface injection, so it's technically extendable.

**Review by OroArmor — APPROVED** [review]

**Comment by TheGlitch76** [comment]
merge conflicts

**Comment by LambdAurora** [comment]
Merge conflict fixed.

**Review by TropheusJ — APPROVED** [review]

**Review by TheGlitch76 — COMMENTED** [review]
I don't really have time to review this again, so in the interest of not holding this PR up i'm dismissing my review.

Per the PR policy, any member of the QSL team can approve this PR (since Oro has already approved), and this is good to go for FCP.

File: library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/QuiltResourcePack.java
Comment (TheGlitch76) [reply library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/QuiltResourcePack.java:29]: Use the annotation please :tiny_potato:

Comment (LambdAurora) [reply library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/QuiltResourcePack.java:29]: Done!

**Review by Leo40Git — APPROVED** [review]

File: library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/QuiltResourcePack.java
Comment (BasiqueEvangelist) [inline library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/QuiltResourcePack.java]: Why is this not `default`?

Comment (LambdAurora) [reply library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/QuiltResourcePack.java]: This is an injected interface, the default implementation is in the mixin.
If a `default` was provided it wouldn't work well.

Comment (BasiqueEvangelist) [reply library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/QuiltResourcePack.java]: Yes, but this means that all mods that have classes that implement ResourcePack will need to also implement QuiltResourcePack, otherwise javac will complain.

Comment (LambdAurora) [reply library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/QuiltResourcePack.java]: Yes.

Not sure what the proper path should be.
Should I add a simple default dummy return value?

**Review by OroArmor — APPROVED** [review]

**Pull request merged**
